src/common/pf_alarm.c

```diff
@@ -2032,6 +2115,177 @@ static int pf_alarm_apmr_periodic (pnet_t * net, pf_ar_t * p_ar)
    return ret;
 }
 
+/**
+ * Send alarm
+ * - Update diag summary
+ * - Update sequence numbers
+ * - Hand over to alarm to APMS instance
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param p_ar             InOut: The AR instance.
+ * @param p_apmx           InOut: The APMX instance.
+ * @param alarm_data       In: Alarm
+ */
+static int pf_alarm_send_internal (
+   pnet_t * net,
+   pf_ar_t * p_ar,
+   pf_apmx_t * p_apmx,
+   pf_alarm_data_t * alarm_data)
+{
+   int ret = -1;
+   uint32_t maint_status = 0;
+
+   CC_ASSERT (net->global_alarm_enable == true);
+   CC_ASSERT (p_ar->alarm_enable == true);
+   CC_ASSERT (p_apmx->p_alpmx->alpmi_state == PF_ALPMI_STATE_W_ALARM);
+
+   (void)pf_alarm_get_diag_summary (
+      net,
+      p_ar,
+      alarm_data->api_id,
+      alarm_data->slot_nbr,
+      alarm_data->subslot_nbr,
+      NULL,
+      &alarm_data->alarm_specifier,
+      &maint_status);
+
+   /* Calculate sequence numbers */
+   alarm_data->sequence_number = p_apmx->p_alpmx->sequence_number;
+   p_apmx->p_alpmx->prev_sequence_number = p_apmx->p_alpmx->sequence_number;
+   p_apmx->p_alpmx->sequence_number++;
+   p_apmx->p_alpmx->sequence_number %= 0x0800;
+
+   /* Create and send an DATA RTA-PDU */
+   ret = pf_alarm_apms_apms_a_data_req (
+      net,
+      p_apmx,
+      alarm_data,
+      maint_status,
+      NULL);
+
+   if (ret == 0)
+   {
+      p_apmx->p_alpmx->alpmi_state = PF_ALPMI_STATE_W_ACK;
+   }
+   else
+   {
+      LOG_ERROR (PF_ALARM_LOG, "Alarm(%d): Failed to send alarm\n", __LINE__);
+   }
+
+   return ret;
+}
+
+/**
+ * Reset queues for outgoing alarms
+ * @param net              InOut: The p-net stack instance
+ * @param p_ar             InOut: The AR instance.
+ * @return 0 is always returned
+ */
+static int pf_alarm_reset_send_queues (pnet_t * net, pf_ar_t * p_ar)
+{
+   memset (p_ar->alarm_send_q, 0, sizeof (p_ar->alarm_send_q));
+   return 0;
+}
+
+/**
+ * Add an alarm to the send queue
+ * @param q               InOut: Alarm send queue
+ * @param p_alarm_data    InOut: Alarm
+ * @return 0  is returned if an alarm is put into the queue
+ *         -1 is returned if queue is full
+ */
+static int pf_alarm_add_send_queue (
+   pf_alarm_queue_t * q,
+   pf_alarm_data_t * p_alarm_data)
+{
+   int ret = -1;
+
+   if (q->count < PNET_MAX_ALARMS)
+   {
+      memcpy (&q->items[q->write_index], p_alarm_data, sizeof (*p_alarm_data));
+      q->write_index++;
+      q->write_index %= PNET_MAX_ALARMS;
+      q->count++;
+      ret = 0;
+   }
+   else
+   {
+      LOG_ERROR (
+         PF_ALARM_LOG,
+         "Alarm(%d): Send queue is full, alarm dropped.\n",
+         __LINE__);
+   }
+
+   return ret;
+}
+
+/**
+ * Fetch an alarm from the send queue
+ * @param q               InOut: Alarm send queue
+ * @param p_alarm_data    InOut: Alarm
```

Comment:
Out (as the initial value doesn't matter)

Reply:
Done